src/libsystemd/sd-device/sd-device.c

```diff
@@ -415,6 +415,36 @@ _public_ int sd_device_new_from_stat_rdev(sd_device **ret, const struct stat *st
         return sd_device_new_from_devnum(ret, type, st->st_rdev);
 }
 
+_public_ int sd_device_new_from_devname(sd_device **ret, const char *devname) {
+        struct stat st;
+
+        assert_return(ret, -EINVAL);
+        assert_return(devname, -EINVAL);
+
+        if (!path_startswith(devname, "/dev"))
+                return -EINVAL;
+
+        if (stat(devname, &st) < 0)
+                return errno == ENOENT ? -ENODEV : -errno;
+
+        return sd_device_new_from_stat_rdev(ret, &st);
+}
+
+_public_ int sd_device_new_from_path(sd_device **ret, const char *path) {
+        assert_return(ret, -EINVAL);
+        assert_return(path, -EINVAL);
+
+        if (path_startswith(path, "/dev"))
+                return sd_device_new_from_devname(ret, path);
+
+        if (path_startswith(path, "/sys"))
+                return sd_device_new_from_syspath(ret, path);
+
+        /* Assume that the path is without "/sys" prefix, like the path obtained by sd_device_get_devpath(). */
+        path = strjoina("/sys", path);
```

Comment:
hmm, so I am not conviced by this fallback, since it makes it a lot harder to extend this function with different kinds of paths.

I mean, I think the fallback makes sense in a user-facing tool, because users want to be quicky and sloppy, but it does create some conceptual ambiguity as /dev/block/7:0 could in theory mean two things with such a fallback: actually /dev/block/7:0, the device node, or /sys/dev/block/7:0 which also exists! (really, it does, it's loop0). The ambiguity is only suppressed because this code prefers understanding of /dev over /sys, but I think that's still ugly.

So I'd just remove the ambiguity as part of the sd-device. i.e. if the path starts neither with /dev/ nor with /sys/ I'd just return an error, and thus avoid any ambiguity, and keeping things perfectly extensible if we should ever need to add more stuff here...

(and this should have been path_join(), since otherwise passing "xyz" as argument would have created a path "/sysxyz" here, which definitely is not desirable...)